# Shopify Data Science Intern Challenge - Question One

In [1]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

In [2]:
# load dataset into a pandas dataframe and show all the rows
df = pd.read_csv('shopify-data-science-intern-challenge-data-set.csv', index_col = False)
pd.set_option('display.max_rows', None)

In [3]:
# high-level overview of data shape and composition
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   order_id        5000 non-null   int64 
 1   shop_id         5000 non-null   int64 
 2   user_id         5000 non-null   int64 
 3   order_amount    5000 non-null   int64 
 4   total_items     5000 non-null   int64 
 5   payment_method  5000 non-null   object
 6   created_at      5000 non-null   object
dtypes: int64(5), object(2)
memory usage: 273.6+ KB


In [4]:
# check for duplicate rows
df.duplicated().any()

False

In [5]:
# show the first ten rows of the dataframe
df.head(10)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   
5         6       58      882           138            1    credit_card   
6         7       87      915           149            1           cash   
7         8       22      761           292            2           cash   
8         9       64      914           266            2          debit   
9        10       52      788           146            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11  
5  2017-03-14 15:25:01  
6  2017-03-01 21:37:57  
7   2017-03-08 2:05:38  
8  2017-03-17 20:56:50  
9  2017-03-30 21:08:26

The Average Order Value (AOV) that was orginally calculated was $3145.13. However, since we know that the shops are selling sneakers, a comparitively affordable item, this value seems to be alot more than expected. To look into what might be wrong with our calculation, let's dig deeper into the variables that might affect our AOV. The order_amount and total_items variables seem like a good place to start,

In [6]:
# descriptive stats for quant variables of interest
df[['order_amount','total_items']].describe()

order_amount  total_items
count    5000.000000   5000.00000
mean     3145.128000      8.78720
std     41282.539349    116.32032
min        90.000000      1.00000
25%       163.000000      1.00000
50%       284.000000      2.00000
75%       390.000000      3.00000
max    704000.000000   2000.00000

It looks like the AOV of $3145.13 that was originally calculated was based on the mean order_amount of all the orders in the data set.

However, the max order_amount of $704,000 and max total_items of 2,000 are really high compared to the values at the 75th percentile. Let's look at the 98th percentile,

In [7]:
np.percentile(df['order_amount'], 98), np.percentile(df['total_items'], 98)

(800.0799999999981, 5.0)

The values above are still reasonable to what we can expect for typical orders. The max order_amount of $704,000 and max total_items of 2000 are definitely outliers. Let's see if there are more orders like these,

In [8]:
df[df['order_amount'] > 801]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
160        161       78      990         25725            1    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
511        512       78      967         51450            2           cash   
520        521       42      607        704000         2000    credit_card   
522        523       46      761           830            5    credit_card   
617        618       78      760         51450            2           cash   
691        692       78      878        154350            6          debit   
742        743       12      727           804            4           cash   
771        772       19      818           815            5          debit   
879        880       60      870           885            5          debit   
938        939       42      808          1056            3    credit_card   
1056      1057       78      800         25725            1          debit   
1104      1105       42      607        704000         2000    credit_card   
1123      1124       29      911           815            5    credit_card   
1150      1151       82      853           885            5          debit   
1193      1194       78      944         25725            1          debit   
1204      1205       78      970         25725            1    credit_card   
1256      1257        6      942           935            5    credit_card   
1259      1260       78      775         77175            3    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1364      1365       42      797          1760            5           cash   
1367      1368       42      926          1408            4           cash   
1384      1385       78      867         25725            1           cash   
1419      1420       78      912         25725            1           cash   
1436      1437       42      607        704000         2000    credit_card   
1452      1453       78      812         25725            1    credit_card   
1471      1472       42      907          1408            4          debit   
1529      1530       78      810         51450            2           cash   
1562      1563       42      607        704000         2000    credit_card   
1563      1564       91      934           960            6          debit   
1602      1603       42      607        704000         2000    credit_card   
1764      1765       12      789           804            4          debit   
1946      1947       33      866           865            5           cash   
1962      1963       46      879           830            5          debit   
2032      2033       88      798           880            5           cash   
2039      2040       11      756           920            5    credit_card   
2153      2154       42      607        704000         2000    credit_card   
2270      2271       78      855         25725            1    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2307      2308       61      723           948            6    credit_card   
2353      2354       27      811           845            5           cash   
2452      2453       78      709         51450            2           cash   
2492      2493       78      834        102900            4          debit   
2494      2495       50      757           965            5          debit   
2495      2496       78      707         51450            2           cash   
2512      2513       78      935         51450            2          debit   
2548      2549       78      861         25725            1           cash   
2560      2561        6     

In [9]:
df[df['total_items'] > 5]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
691        692       78      878        154350            6          debit   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
1563      1564       91      934           960            6          debit   
1602      1603       42      607        704000         2000    credit_card   
2127      2128       83      745           774            6           cash   
2153      2154       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2307      2308       61      723           948            6    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
3252      3253       67      706           786            6    credit_card   
3332      3333       42      607        704000         2000    credit_card   
3538      3539       43      830          1086            6          debit   
3865      3866       68      815           816            6          debit   
4056      4057       42      607        704000         2000    credit_card   
4141      4142       54      733          1064            8          debit   
4646      4647       42      607        704000         2000    credit_card   
4711      4712       86      883           780            6           cash   
4847      4848       13      993           960            6           cash   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   

               created_at  
15     2017-03-07 4:00:00  
60     2017-03-04 4:00:00  
520    2017-03-02 4:00:00  
691   2017-03-27 22:51:43  
1104   2017-03-24 4:00:00  
1362   2017-03-15 4:00:00  
1436   2017-03-11 4:00:00  
1562   2017-03-19 4:00:00  
1563   2017-03-23 8:25:49  
1602   2017-03-17 4:00:00  
2127   2017-03-27 6:59:46  
2153   2017-03-12 4:00:00  
2297   2017-03-07 4:00:00  
2307  2017-03-26 11:29:37  
2835   2017-03-28 4:00:00  
2969   2017-03-28 4:00:00  
3252  2017-03-29 16:05:41  
3332   2017-03-24 4:00:00  
3538  2017-03-17 19:56:29  
3865   2017-03-11 9:31:50  
4056   2017-03-28 4:00:00  
4141  2017-03-07 17:05:18  
4646   2017-03-02 4:00:00  
4711  2017-03-18 14:18:19  
4847  2017-03-27 11:00:45  
4868   2017-03-22 4:00:00  
4882   2017-03-25 4:00:00

In [10]:
df[df['shop_id'] == 42]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
40          41       42      793           352            1    credit_card   
60          61       42      607        704000         2000    credit_card   
308        309       42      770           352            1    credit_card   
409        410       42      904           704            2    credit_card   
520        521       42      607        704000         2000    credit_card   
834        835       42      792           352            1           cash   
835        836       42      819           704            2           cash   
938        939       42      808          1056            3    credit_card   
979        980       42      744           352            1          debit   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1364      1365       42      797          1760            5           cash   
1367      1368       42      926          1408            4           cash   
1436      1437       42      607        704000         2000    credit_card   
1471      1472       42      907          1408            4          debit   
1512      1513       42      946           352            1          debit   
1520      1521       42      756           704            2          debit   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
1911      1912       42      739           704            2           cash   
1929      1930       42      770           352            1    credit_card   
2003      2004       42      934           704            2           cash   
2018      2019       42      739           352            1          debit   
2053      2054       42      951           352            1          debit   
2153      2154       42      607        704000         2000    credit_card   
2273      2274       42      747           704            2          debit   
2297      2298       42      607        704000         2000    credit_card   
2491      2492       42      868           704            2          debit   
2609      2610       42      868           704            2          debit   
2766      2767       42      970           704            2    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
2987      2988       42      819          1056            3           cash   
3332      3333       42      607        704000         2000    credit_card   
3513      3514       42      726          1056            3          debit   
3651      3652       42      830           352            1    credit_card   
3697      3698       42      839           352            1          debit   
3903      3904       42      975           352            1          debit   
3998      3999       42      886           352            1          debit   
4056      4057       42      607        704000         2000    credit_card   
4231      4232       42      962           352            1           cash   
4294      4295       42      859           704            2           cash   
4326      4327       42      788           704            2          debit   
4421      4422       42      736           704            2    credit_card   
4625      4626       42      809           352            1    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4745      4746       42      872           352            1          debit   
4767      4768       42      720           704            2    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42     

In [11]:
df[df['user_id'] == 607]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
2153      2154       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
3332      3333       42      607        704000         2000    credit_card   
4056      4057       42      607        704000         2000    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   

              created_at  
15    2017-03-07 4:00:00  
60    2017-03-04 4:00:00  
520   2017-03-02 4:00:00  
1104  2017-03-24 4:00:00  
1362  2017-03-15 4:00:00  
1436  2017-03-11 4:00:00  
1562  2017-03-19 4:00:00  
1602  2017-03-17 4:00:00  
2153  2017-03-12 4:00:00  
2297  2017-03-07 4:00:00  
2835  2017-03-28 4:00:00  
2969  2017-03-28 4:00:00  
3332  2017-03-24 4:00:00  
4056  2017-03-28 4:00:00  
4646  2017-03-02 4:00:00  
4868  2017-03-22 4:00:00  
4882  2017-03-25 4:00:00

In [12]:
df[df['shop_id'] == 78]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
160        161       78      990         25725            1    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
511        512       78      967         51450            2           cash   
617        618       78      760         51450            2           cash   
691        692       78      878        154350            6          debit   
1056      1057       78      800         25725            1          debit   
1193      1194       78      944         25725            1          debit   
1204      1205       78      970         25725            1    credit_card   
1259      1260       78      775         77175            3    credit_card   
1384      1385       78      867         25725            1           cash   
1419      1420       78      912         25725            1           cash   
1452      1453       78      812         25725            1    credit_card   
1529      1530       78      810         51450            2           cash   
2270      2271       78      855         25725            1    credit_card   
2452      2453       78      709         51450            2           cash   
2492      2493       78      834        102900            4          debit   
2495      2496       78      707         51450            2           cash   
2512      2513       78      935         51450            2          debit   
2548      2549       78      861         25725            1           cash   
2564      2565       78      915         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2773      2774       78      890         25725            1           cash   
2818      2819       78      869         51450            2          debit   
2821      2822       78      814         51450            2           cash   
2906      2907       78      817         77175            3          debit   
2922      2923       78      740         25725            1          debit   
3085      3086       78      910         25725            1           cash   
3101      3102       78      855         51450            2    credit_card   
3151      3152       78      745         25725            1    credit_card   
3167      3168       78      927         51450            2           cash   
3403      3404       78      928         77175            3          debit   
3440      3441       78      982         25725            1          debit   
3705      3706       78      828         51450            2    credit_card   
3724      3725       78      766         77175            3    credit_card   
3780      3781       78      889         25725            1           cash   
4040      4041       78      852         25725            1           cash   
4079      4080       78      946         51450            2           cash   
4192      4193       78      787         77175            3    credit_card   
4311      4312       78      960         51450            2          debit   
4412      4413       78      756         51450            2          debit   
4420      4421       78      969         77175            3          debit   
4505      4506       78      866         25725            1          debit   
4584      4585       78      997         25725            1           cash   
4715      4716       78      818         77175            3          debit   
4918      4919       78      823         25725            1           cash   

               created_at  
160    2017-03-12 5:56:57  
490   2017-03-26 17:08:19  
493   2017-03-16 21:39:35  
511    2017-03-09 7:23:14  
617   2017-03-18 11:18:42  
691   2017-03-27 22:51:43  
1056  2017-03-15 10:16:45  
1193  2017-03-16 16:38:26  
1204  2017-03-17 22:32:21  
1259   2017-03-27 9:27:20  
1384  2017-03-17 16:38:06  
141

We can see that there are alot of wholesale orders that have high order_amounts. This is what's causing the AOV to be much higher than expected. 

In particular, there's two stores that stand out. The store with shop_id 42 has orders that have large order_amounts and large total_items. Most of these orders are associated with user_id 607. In addition to this, the store with shop_id 78 has orders that have large order_amounts and only a few total_items. This leads me to believe the orders for shop_id 78 either have incorrect order_amounts or incorrect total_items.

We can solve these issues by cleaning our data set. 98% of our orders are under approx. $800 and the wholesale orders with shop_id 42 and user_id 607 are outliers and not reflective of typical orders. Furthermore, the orders for shop_id 78 are erroneous. So we can safely ignore these orders in calculating our AOV,

In [13]:
# drop all the rows that are associated with user_id 607
df2 = df[df['user_id'] != 607]

In [14]:
# drop all the rows that are associated with shop_id 78
df3 = df2[df2['shop_id'] != 78]

In [15]:
# descriptive stats for cleaned dataset
df3[['order_amount','total_items']].describe()

order_amount  total_items
count   4937.000000  4937.000000
mean     302.580514     1.994734
std      160.804912     0.982821
min       90.000000     1.000000
25%      163.000000     1.000000
50%      284.000000     2.000000
75%      387.000000     3.000000
max     1760.000000     8.000000

In [16]:
# 95th percentile of the order_amount for the cleaned dataset
np.percentile(df3['order_amount'], 95)

592.0

The mean AOV of 302.58 seems much more reasonable than the AOV of 3145.13 that was calculated earlier. However it might be best to use the median order value of 284.00 to be safe, as it is not affected by the few larger orders that are present.

If we were not looking for a single value, I'd group all the orders by store and then look at the median order value for each store as below,

In [17]:
# create groupby object to group all the orders by each store
group_by_store = df3.groupby(['shop_id'])

In [18]:
# descriptive stats for the order_amount and total_items for each store
group_by_store['order_amount'].agg([np.mean, np.median, np.std, np.min, np.max])

mean  median         std  amin  amax
shop_id                                            
1        308.818182     316  117.816032   158   632
2        174.327273     188   91.204909    94   470
3        305.250000     296  141.253130   148   592
4        258.509804     256  103.956793   128   512
5        290.311111     284  148.176771   142   710
6        383.508475     374  188.361206   187   935
7        218.000000     224  113.857327   112   560
8        241.043478     198  134.338702   132   660
9        234.000000     236  120.996438   118   590
10       332.301887     296  156.050976   148   740
11       356.734694     368  189.321021   184   920
12       352.698113     402  166.803268   201   804
13       345.396825     320  157.938590   160   960
14       242.000000     232  129.070632   116   580
15       308.942308     306  152.971151   153   765
16       270.146341     312  139.636951   156   780
17       332.075472     352  167.894638   176   704
18       342.588235     312  187.250973   156   780
19       320.906250     326  164.208319   163   815
20       251.557692     254  131.863330   127   508
21       308.695652     284  138.175552   142   710
22       273.750000     292  107.012527   146   584
23       317.672727     312  146.966745   156   780
24       320.727273     280  110.014385   140   560
25       232.916667     260  110.433543   130   520
26       341.224490     352  184.618894   176   880
27       334.870370     338  159.115367   169   845
28       320.372093     328  133.682820   164   656
29       331.620690     326  152.558166   163   815
30       295.071429     306  139.487746   153   612
31       268.978723     258  113.579218   129   516
32       189.976190     202   81.053276   101   404
33       376.275000     346  187.425197   173   865
34       234.240000     244  120.345516   122   488
35       328.000000     328  152.329879   164   656
36       254.800000     260  117.338513   130   650
37       340.208333     284  184.657124   142   710
38       390.857143     380  189.680403   190   760
39       268.000000     268  137.309140   134   670
40       295.166667     322  130.049963   161   644
41       254.000000     236  136.513230   118   590
42       652.235294     704  358.681682   352  1760
43       333.913793     362  185.087394   181  1086
44       262.153846     288  131.612977   144   576
45       269.310345     284  163.296545   142   710
46       347.441860     332  197.791774   166   830
47       259.148936     145  141.606924   145   725
48       242.775000     234  113.613490   117   468
49       279.905660     258  153.337394   129   645
50       403.545455     386  185.297681   193   965
51       361.804348     374  190.712188   187   935
52       316.926829     292  162.914148   146   730
53       214.117647     224  101.908875   112   560
54       276.640000     266  171.716278   133  1064
55       327.750000     342  148.963011   171   684
56       218.189189     234  114.244581   117   468
57       296.773585     294  139.726056   147   735
58       254.949153     276  127.651117   138   552
59       358.966667     356  180.967079   178   712
60       350.234043     354  162.932663   177   885
61       344.440000     316  193.351658   158   948
62       308.837209     320  149.749144   160   640
63       264.965517     272  112.271126   136   544
64       272.186047     266  108.413506   133   399
65       330.814815     308  153.730722   154   616
66       312.886792     322  182.520037   161   805
67       272.621622     262  164.485959   131   786
68       254.638298     272  143.528883   136   816
69       264.183333     262  121.775355   131   655
70       343.067797     346  171.476636   173   865
71       323.030303     328  131.734111   164   656
72       309.565217     320  144.698253   160   640
73       335.689655     330  134.599628   165   660
74       306.000000     306  155.053783   153   765
75       240.761905     256  110.227218   128   512
76       32

## Summary

Q1. a) Think about what could be going wrong with our calculation. Think about a better way to evaluate this data.

Answer: There are some wholesale orders that have high order_amounts. This is what's causing the AOV to be much higher than expected. 

In particular, there's two stores that stand out. The store with shop_id 42 has orders that have large order_amounts and large total_items. Most of these orders are associated with user_id 607. In addition to this, the store with shop_id 78 has orders that have large order_amounts and only a few total_items. This leads me to believe the orders for shop_id 78 either have incorrect order_amounts or incorrect total_items.

We can solve these issues by cleaning our data set. 98% of our orders are under approx. $800 and the wholesale orders with shop_id 42 and user_id 607 are outliers and not reflective of typical orders. Furthermore, the orders for shop_id 78 are erroneous. So we can safely ignore these orders in calculating our AOV.

Q1. b) What metric would you report for this dataset?

Answer: The mean AOV of 302.58 from the cleaneed data set seems much more reasonable than the AOV of 3145.13 that was calculated earlier. However it might be best to use the median order value of 284.00 to be safe, as it is not affected by the few larger orders that are present.

If we were not looking for a single value, I’d ignore the wholesale and incorrect orders, group the orders by store and then find the median order value for each store.

Q1. c) What is its value?

Answer: The mean order value is $284.00.